### PR TITLE
The number of tests in Belgium from Belgium

### PR DESCRIPTION
**Pull request by cast42**
I corrected the error: pd -> pandas

### CODE REVIEW

**Review by AlexAndorra — COMMENTED** [review]
Hi @cast42, and thanks for the PR!
This looks like a good start. I noticed the regions are commented out, so I'm guessing this is still WIP, right?
Also, I think it'll be useful to add a dict of short names for the regions. As is, the regions' codes will be displayed in the plot (1, 2, etc.), which I guess is not often used in Belgium? You can take a look at such a dict in `data_fr` or `data_it` -- that way, plots will display a more intuitive name for the regions 😉

**Comment by cast42** [comment]
Yes, WIP. Regions are already ok. Still have to do provinces. I'll add short names asap. I'll use english names as currently it not foreseen to have two languages per country.

**Comment by cast42** [comment]
I think you might review what I produced so far. Data is aggregated at regional and province level.

**Comment by cast42** [comment]
Apologies for the many commits. I was not able to set up the environment due to issues with qt:
With conda 4.9.2 on Windows 10 I get:
Executing transaction: done
ERROR conda.core.link:_execute(700): An error occurred while installing package 'conda-forge::qt-5.12.9-hb2cf2c5_0'.
I tried to remove python-graphviz package from environment.yml because I guestimated that it required the qt library but I got the same error.

I managed to get a colab notebook up and running:
https://colab.research.google.com/drive/1GzqcQywuNlbiUYj6GSSLhbsK6doMMxFe?usp=sharing
At least the preprocessing seems to work!

**Comment by michaelosthege** [comment]
This looks great! I will check out the branch and run it with the model to verify.

In the meantime, if you like you can try out these install commands for Pymc3 on Windows: https://gist.github.com/michaelosthege/5bd75c99cd5e806ee049b02ed528bab3

Also, just a question out of curiosity: the latest numbers for Belgium show a sharp decline in the number of total tests: Do you know why? Did they run out of reagents or something?

**Comment by cast42** [comment]
Yes, the testing strategy in Belgium was changed because it could not cope with the demand: https://www.vrt.be/vrtnws/en/2020/10/20/health-minister-on-coronavirus-testing-it-needs-to-be-kept-feasi/.
Asymtomatic and people returning from a red zone without symptoms are not tested anymore for the time being.
In the mean time, testing capacity is increasing again: https://www.vrt.be/vrtnws/en/2020/11/02/first-three-national-coronavirus-testing-super-labs-open-in-ghen/ It is expected that the testing capacity will be doubled by the end of the year. Maybe we should add a change point to Prophet 😀

I will look into installation issues next week. But it might be an issue with permissions. (No admin rights...)

Now that I have looked into the code, I want to say the following: I’m impressed by it’s quality. Great work! Congratulations to the team, it is clearly high skilled.

Looking forward to lots of Re curves.

L

File: rtlive/sources/data_be.py [hunk 1]
```diff
@@ -1,39 +1,225 @@
...
+
+from typing import Dict, Tuple, Union
 
-from . import ourworldindata
 from .. import preprocessing
 
 _log = logging.getLogger(__file__)
 
+# From https://en.wikipedia.org/wiki/Provinces_of_Belgium
+BE_REGION_NAMES = {
```
Comment (michaelosthege) [inline rtlive/sources/data_be.py:14]: The data workflow uses the region short codes (in this case 3-letter alphabetic) in all places.
This means we'll need these dicts to use the short codes as keys.

File: rtlive/sources/data_be.py [hunk 2]
```diff
@@ -1,39 +1,225 @@
...
+    'Antwerpen': 'ANT',
+    'Limburg': 'LIM',
+    'OostVlaanderen': 'EFL',
+    'VlaamsBrabant': 'FBR',
+    'WestVlaanderen': 'WFL',
+    'Hainaut': 'HAI',
+    'Liège': 'LIE',
+    'Luxembourg': 'LUX',
+    'Namur': 'NAM',
+    'BrabantWallon': 'WBR',
```
Comment (michaelosthege) [inline rtlive/sources/data_be.py:48]: This dict is optional: You can use it to map short codes (FLA, ANT, ...) to more commonly known abbreviations.
For most countries you don't need it.

File: rtlive/sources/data_be.py [hunk 3]
```diff
@@ -1,39 +1,225 @@
...
-    """ Applies testcount interpolation/extrapolation.
+BE_REGION_CODES = {
+    v : k
+    for k, v in BE_REGION_NAMES.items()
+}
 
-    Currently this assumes the OWID data, which only has an "all" region.
-    In the future, this should be replaced with more fine graned data loading!
+# Source: https://www.ibz.rrn.fgov.be/fileadmin/user_upload/fr/pop/statistiques/population-bevolking-20200101.pdf
+BE_REGION_POPULATION = {
```
Comment (michaelosthege) [inline rtlive/sources/data_be.py:57]: also here: short codes as keys

File: rtlive/sources/data_be.py [hunk 4]
```diff
@@ -1,39 +1,225 @@
...
+    -------
+    df : pandas.DataFrame
+        table with columns as required by rtlive/data.py API
+    """
+    
+    def redistribute(g, col):
+        gdata = g.groupby('REGION')[col].sum()
+        gdata.loc['Brussels'] += gdata.loc['Nan'] * (gdata.loc['Brussels']/(gdata.loc['Brussels'] + gdata.loc['Flanders'] + gdata.loc['Wallonia']))
+        gdata.loc['Flanders'] += gdata.loc['Nan'] * (gdata.loc['Flanders']/(gdata.loc['Brussels'] + gdata.loc['Flanders'] + gdata.loc['Wallonia']))
+        gdata.loc['Wallonia'] += gdata.loc['Nan'] * (gdata.loc['Wallonia']/(gdata.loc['Brussels'] + gdata.loc['Flanders'] + gdata.loc['Wallonia']))
```
Comment (michaelosthege) [inline rtlive/sources/data_be.py:92]: Also here we use the short codes on the index, so you'll have to replace them. You can use the `BE_REGION_CODES` lookup for that.

File: rtlive/sources/data_be.py [hunk 5]
```diff
@@ -1,39 +1,225 @@
...
     compute_zone=data.Zone.Europe,
-    region_name={
-        "all": "Belgium",
-    },
-    region_population={
-        "all": 11_492_641,
-    },
-    fn_load=ourworldindata.create_loader_function("BE"),
+    region_name=BE_REGION_NAMES,
+    region_short_name=BE_REGION_ABBR,
```
Comment (michaelosthege) [inline rtlive/sources/data_be.py]: abbreviations are optional and in this case not needed

File: rtlive/sources/data_be.py [hunk 6]
```diff
@@ -1,39 +1,225 @@
...
+        raise NotImplementedError(
+            'Downloading with a run_date is not yet supported. '
+            f'Today: {datetime.date.today()}, run_date: {run_date}'
+        )
+        
+    # Download data from Sciensano and remove last two days to deal with reporting delays
+    df_tests = (pandas.read_csv('https://epistat.sciensano.be/Data/COVID19BE_tests.csv', parse_dates=['DATE'])
+       .set_index('DATE')
+       .sort_index()
+       .truncate(after=run_date - pandas.DateOffset(2))
```
Comment (michaelosthege) [inline rtlive/sources/data_be.py]: Depending on which reporting delay this truncation is about, the model may already account for it*!

*This is done through the "delay distribution" (see cell 9 of [this notebook](https://github.com/rtcovidlive/rtlive-global/blob/master/notebooks/Tutorial_model.ipynb)). Ideally, we'd need to calculate this empirical distribution differently for all countries, but due to lack of time/data this is not done.

**Review by michaelosthege — CHANGES_REQUESTED** [review]
I tried to change the short code stuff in the dataframe, but was confused that there are "province" and "region", so it's better if you make this change.

**Comment by michaelosthege** [comment]
> Yes, the testing strategy in Belgium was changed because it could not cope with the demand: https://www.vrt.be/vrtnws/en/2020/10/20/health-minister-on-coronavirus-testing-it-needs-to-be-kept-feasi/.
> Asymtomatic and people returning from a red zone without symptoms are not tested anymore for the time being.
> In the mean time, testing capacity is increasing again: https://www.vrt.be/vrtnws/en/2020/11/02/first-three-national-coronavirus-testing-super-labs-open-in-ghen/ It is expected that the testing capacity will be doubled by the end of the year. Maybe we should add a change point to Prophet 😀

Thanks for that link!! And good to see such efforts. I work in lab automation myself and such high-throughput efforts are just really cool. Unfortunately it doesn't look like Germany does this anywhere.

> I will look into installation issues next week. But it might be an issue with permissions. (No admin rights...)

Never install any Python stuff (on Windows) with admin permissions. Start with non-admin installation of miniconda and go from there. I learned that the hard way..

> Now that I have looked into the code, I want to say the following: I’m impressed by it’s quality. Great work! Congratulations to the team, it is clearly high skilled.

Thank you!

**Review by cast42 — COMMENTED** [review]
I update the code that uses short codes everywhere.

File: rtlive/sources/data_be.py (hunk 6 above)
Comment (cast42) [reply rtlive/sources/data_be.py]: Nifty, I removed the truncate from the input timeline

**Comment by michaelosthege** [comment]
@cast42 When I run the dataloading notebook with your updated branch, I still get this error:
grafik

**Comment by cast42** [comment]
I guess tha is because the old code is still there.
Execute !rm -fR rtlive-global in a cell and then restart the kernel and run all.
The git clone in the first cell will then reload the latest version of the code.
Good luck. Let me know if there are still issues.

**Comment by cast42** [comment]
Strange. It looks like the csv file is not read. Can you try to execute the following code cells in your notebook:
   ```
df_tests = (pandas.read_csv('https://epistat.sciensano.be/Data/COVID19BE_tests.csv', parse_dates=['DATE'])
       .set_index('DATE')
       .sort_index()
       .reset_index()
    )
df_tests.head()
```
   ```
# Reformat data into Rtlive.de format at country level all
df_tests_per_all_day = (df_tests
       .assign(region='all')
       .groupby('DATE', as_index=False)
       .agg(new_cases=('TESTS_ALL_POS', 'sum'), new_tests=('TESTS_ALL', 'sum'), region=('region', 'first'))
       .rename(columns={'DATE':'date'})
       .set_index(['region', "date"])
       .sort_index()
    )
```
If the above cell does not give an error, inspect first lines of df_tests_per_all_day dataframe:
`df_tests_per_all_day.head()`

Please also verify the version of pandas with `pandas.__version__`

In this notebook https://colab.research.google.com/drive/1GzqcQywuNlbiUYj6GSSLhbsK6doMMxFe?usp=sharing
it is 1.1.4

Thx,
Lode

**Comment by davideferre** [comment]
@cast42 I've made this simple snippet, to show you how to fix your code:
```python
import pandas
import requests
import io

content = requests.get('https://epistat.sciensano.be/Data/COVID19BE_tests.csv', verify=False,).content
df_tests = pandas.read_csv(
        io.StringIO(content.decode("utf-8")),
        sep=",",
        parse_dates=["DATE"],
        usecols=["DATE", "REGION", "TESTS_ALL_POS", "TESTS_ALL"]
).rename(
	columns={
    	"DATE": "date",
        "REGION": "region",
        "TESTS_ALL_POS": "new_cases",
        "TESTS_ALL": "tests"
    }
)
print(df_tests)
```

Your main problem is that ssl certificate of `epistat.sciensano.be`is not valid. With this code I can fetch csv data.

**Comment by michaelosthege** [comment]
@cast42 Indeed, I was on pandas<1.1.
I just changed the order of column renaming / index setting a bit and now it works.
The results look fine for all regions, so I'll trigger the model workflow now.

But please take a close look at my changes to makes sure that I did not break the grouping/aggregation.

**Comment by cast42** [comment]
This is the resulting [data loading notebook for BE](https://github.com/cast42/rtlive-global/blob/master/notebooks/Tutorial_dataloading.ipynb).

To run for yourself, you can use binder:

Binder](https://mybinder.org/v2/gh/cast42/rtlive-global.git/HEAD?filepath=notebooks%2FTutorial_dataloading.ipynb)

**Comment by cast42** [comment]
This is the resulting [data loading notebook for BE](https://github.com/cast42/rtlive-global/blob/master/notebooks/Tutorial_dataloading.ipynb).

To run for yourself, you can use binder:

[Binder](https://mybinder.org/v2/gh/cast42/rtlive-global.git/HEAD?filepath=notebooks%2FTutorial_dataloading.ipynb)

**Comment by michaelosthege** [comment]
Everything looks good now.
The pipeline ran through and you can check out the results on https://rtlive.de/global.html

@cast42 Let me know if you notice any glitches. Otherwise I'll merge this tomorrow 👍

**Review by michaelosthege — APPROVED** [review]
I updated the notebook - the text still mentioned Switzerland.

**Comment by cast42** [comment]
It's looking great! Thanks for your work. Just one minor detail. The short code for Belgium is  𝜙 in the graph. Maybe it could be set to BE ?

Also good to know is that testing of asymptomatic people that had a high risc contact will resume at the end of next week in Belgium. Hence the number of tests will increase.

**Comment by michaelosthege** [comment]
> The short code for Belgium is 𝜙 in the graph. Maybe it could be set to BE ?

Now that you mention it, it looks like we're not super consistent there. Personally, I prefer the symbol, because it's a little easier to spot.
The dashboard is open source at https://github.com/michaelosthege/rtlive-dash-de

**Pull request merged**